File: Samples/BackgroundTask/InProc BackgroundTask/cs-winui/BackgroundTaskBuilder/BackgroundTask.cs
// Copyright (c) Microsoft Corporation and Contributors.
// Licensed under the MIT License.


using System;
using System.IO; // Path
using System.Collections.Generic; // Queue
using Windows.System.Threading; // ThreadPoolTimer
using System.Runtime.InteropServices; // Guid, RegistrationServices
using Windows.ApplicationModel.Background; // IBackgroundTask
using Windows.UI.Notifications; // ToastNotificationManager, ToastTemplateType, ToastNotification

namespace BackgroundTaskBuilder
{
    // Background task implementation.
    // {87654321-1234-1234-1234-1234567890AE} is GUID to register this task with BackgroundTaskBuilder. Generate a random GUID before implementing.
    [Guid("87654321-1234-1234-1234-1234567890AA")]
    public partial class BackgroundTask : IBackgroundTask
    {
        /// <summary>
        /// This method is the main entry point for the background task. The system will believe this background task
        /// is complete when this method returns.
        /// </summary>
        [MTAThread]
        public void Run(IBackgroundTaskInstance taskInstance)
        {
            // Get deferral to indicate not to kill the background task process as soon as the Run method returns
            _deferral = taskInstance.GetDeferral();
            // Wire the cancellation handler.
            taskInstance.Canceled += this.OnCanceled;

            // Set the progress to indicate this task has started
            taskInstance.Progress = 0;

            _periodicTimer = ThreadPoolTimer.CreatePeriodicTimer(new TimerElapsedHandler(PeriodicTimerCallback), TimeSpan.FromSeconds(1));
        }

        // Simulate the background task activity.
        private void PeriodicTimerCallback(ThreadPoolTimer timer)
        {
            if ((_cancelRequested == false) && (_progress < 100))
            {
                _progress += 10;
            }
            else
            {
                if (_cancelRequested) _progress = -1;
                if (_periodicTimer != null) _periodicTimer.Cancel();

                // Indicate that the background task has completed.
                if (_deferral != null) _deferral.Complete();
            }

            BackgroundTaskBuilder.MainWindow.taskStatus(_progress);
        }

        /// <summary>
        /// This method is signaled when the system requests the background task be canceled. This method will signal
        /// to the Run method to clean up and return.
        /// </summary>
        [MTAThread]
        public void OnCanceled(IBackgroundTaskInstance taskInstance, BackgroundTaskCancellationReason cancellationReason)
        {
            // Handle cancellation operations and flag the task to end
            _cancelRequested = true;
        }

        private BackgroundTaskDeferral? _deferral = null;
        private volatile bool _cancelRequested = false;
        private ThreadPoolTimer? _periodicTimer = null;
        private int _progress = 0;
    }
}

File: Samples/BackgroundTask/InProc BackgroundTask/cpp-winui/BackgroundTaskBuilder/BackgroundTask.cpp
// Copyright (c) Microsoft Corporation and Contributors.
// Licensed under the MIT License.

#include "pch.h"
#include "BackgroundTask.h"
#include "RegisterForCOM.h"
#include "winrt/Windows.Data.Xml.Dom.h"
#include "winrt/Windows.UI.Notifications.h"
#include "App.xaml.h"

using namespace winrt;
namespace winrt
{
    using namespace winrt::Windows::ApplicationModel::Background;
    using namespace Windows::Data::Xml::Dom;
    using namespace Windows::UI::Notifications;
}

namespace winrt::BackgroundTaskBuilder
{
    void BackgroundTask::Run(_In_ IBackgroundTaskInstance taskInstance)
    {
        // Get deferral to indicate not to kill the background task process as soon as the Run method returns
        m_deferral = taskInstance.GetDeferral();
        m_progress = 0;
        taskInstance.Canceled({ this, &BackgroundTask::OnCanceled });

        // Calling a method on the Window to inform that the background task is executed
        winrt::Microsoft::UI::Xaml::Window window = winrt::BackgroundTaskBuilder::implementation::App::Window();
        m_mainWindow = window.as<winrt::BackgroundTaskBuilder::IMainWindow>();

        Windows::Foundation::TimeSpan period{ std::chrono::seconds{2} };
        m_periodicTimer = Windows::System::Threading::ThreadPoolTimer::CreatePeriodicTimer([this, lifetime = get_strong()](Windows::System::Threading::ThreadPoolTimer timer)
            {
                if (!m_cancelRequested && m_progress < 100)
                {
                    m_progress += 10;
                }
                else
                {
                    m_periodicTimer.Cancel();

                    // Indicate that the background task has completed.
                    m_deferral.Complete();
                    if (m_cancelRequested) m_progress = -1;
                }
                m_mainWindow.BackgroundTaskExecuted(m_progress);
            }, period);
    }

    void BackgroundTask::OnCanceled(_In_ IBackgroundTaskInstance /* taskInstance */, _In_ BackgroundTaskCancellationReason /* cancelReason */)
    {
        m_cancelRequested = true;
    }
}
